src/reuse/header.py

```diff
@@ -144,6 +144,86 @@ def create_header(
     return new_header
 
 
+def comment_has_spdx(comment: str) -> bool:
+    """Check whether comment already contains SPDX info."""
+    try:
+        existing_spdx = extract_spdx_info(comment)
+    except (ExpressionError, ParseError):
+        existing_spdx = (None, None)
+    return existing_spdx[0] or existing_spdx[1]
+
+
+def find_initial_spdx(text: str, style: CommentStyle) -> (bool, str, str):
+    """Strip the first comment block of text and see if it has SPDX info."""
+    try:
+        has_spdx = False
+        initial_comment = style.comment_at_first_character(text)
+        if initial_comment:
+            text = text[len(initial_comment) + 1 :]  # also remove newline
+            _LOGGER.debug("Text starts with a comment block")
+            has_spdx = comment_has_spdx(initial_comment)
+            if has_spdx:
+                _LOGGER.debug("Initial comment block contains spdx")
+                initial_has_spdx = True
+            else:
+                _LOGGER.debug("No spdx in initial comment block")
+            return (has_spdx, initial_comment + "\n", text)
+        else:
+            _LOGGER.debug("No initial comment block")
+    except CommentParseError:
+        _LOGGER.debug("Cannot parse initial comment block")
+    return (False, "", text)
+
+
+def find_new_header_place(text: str, style: CommentStyle) -> (str, str, str):
+    """Try to find the best place to put new SPDX info."""
+
+    shebang = None
+    initial_comment = None
+    initial_has_spdx = False
+
+    if text.startswith("#!"):
+        (shebang, text) = text.split("\n", 1)
+        _LOGGER.debug("Found a shebang, removing it temporarily")
+        shebang += "\n"
+
+    _LOGGER.debug("style = {}".format(style))
+
+    (initial_has_spdx, initial_comment, text) = find_initial_spdx(text, style)
+
+    if not initial_has_spdx and shebang:
+        _LOGGER.debug(
+            "Initial block was adjacent to shebang. Trying next one."
```

Comment:
Ah... This bit starts the messy bit, I imagine. I can see a two wrong assumptions here:

- The SPDX comment block cannot be in the same block as a multi-line shebang. (It can be)
- The shebang is necessarily either one or two lines. (It can be really long: <http://rosettacode.org/wiki/Multiline_shebang>)

I initially wrote the code making the following (erroneous) assumptions:

- The shebang is in the same comment block as the SPDX comment.
- The shebang is exactly one line.

The first assumption is *kind of* understandable, because the REUSE Spec mandates that the comment header be the *first comment* in the file. So it would necessarily have to be in the same comment block. But I am not certain whether that is exactly technically feasible with multi-line shebangs.

So ideally, the following things would be handled:

- Identify a single-line shebang.
- If it's a single-line shebang, simply merge the two comment blocks, as has been done thus far.
- Identify a multi-line shebang.
- If it's a multi-line shebang, make the SPDX comment block the *second* comment block in the file. **This causes the program to "break" when people keep the two in the same comment block.**

This kind of forces a certain style, but should be a lot more manageable. The problem is that I'm not entirely certain how to identify a multi-line shebang, given the diversity in the link. You may be on the right path in the code, but more concretely here:

- Parse first comment.
- If it has no SPDX info but *does* have a shebang, then treat *everything* until the next comment block as a shebang.
- Parse that comment block and deal with it.

I'd happily restructure the code to do that, but I'd love some feedback first. Maybe I made some totally wrong assumptions about your approach :)

Reply:
> multi-line shebang

I actually don’t think a multi-line shebang has ever been standardised or something, so I didn’t take it into account tbh. The only official thing is a single line shebang and that’s what my version handles (and if someone puts SPDX headers and multiline shebangs together, they probably deserve getting their code messed up anyway). If you think that special-casing multi-line shebangs is a good idea, then it should be pretty straightforward to cut off not a single line starting with `#!` but all of them.

> the REUSE Spec mandates that the comment header be the first comment in the file

Oh, huh, I somehow missed this. I assumed it was ok to have a multi-line shebang then a new line and then an SPDX comment. I somehow implicitly assumed that the REUSE spec makes an exception for shebangs and “top of file” is actually below the shebang – but now I see that my assumptions were wrong. I’m not even sure how I feel about it, because I was always putting SPDX after a new line below a shebang and apparently that does not conform to the spec.

Reply:
I guess I would suggest to alter the specification to say that SPDX has to be in the first comment block in the file, not necessarily at the top for two reasons:

1. This allows to separate it from the shebang by a new line (and I personally believe that not separating it is ugly), if we assume that a shebang is a shebang and not a comment.
2. There _are_ cases where it is just impossible to put a comment at the top. (This happened to myself just a couple of days ago: I copied a file that implements the standard library for a slightly esoteric programming language, so I added a REUSE header to it using the tool and it stopped working, because... the syntax for comments was not built into the language but rather defined as the first thing in this exact file. So to copy this file to my project with an appropriate header I had to leave the comment syntax definition as the first line of this stdlib and add a REUSE header right after it – as the first _comment_ but not first _thing_ in the file.)